pulp_ansible/app/tasks/signature.py

```diff
@@ -0,0 +1,114 @@
+import asyncio
+import json
+import tempfile
+import hashlib
+import gettext
+
+from pulpcore.plugin.stages import (
+    ContentSaver,
+    DeclarativeContent,
+    DeclarativeVersion,
+    Stage,
+)
+
+from pulp_ansible.app.models import (
+    AnsibleRepository,
+    CollectionVersion,
+    CollectionVersionSignature,
+)
+
+from pulpcore.plugin.models import SigningService, ProgressReport
+from pulpcore.plugin.sync import sync_to_async_iterable, sync_to_async
+
+_ = gettext.gettext
+
+
+def sign(repository_href, content_hrefs, signing_service_href):
+    """The signing task."""
+    repository = AnsibleRepository.objects.get(pk=repository_href)
+    if content_hrefs == ["*"]:
+        filtered = repository.latest_version().content.filter(
+            pulp_type=CollectionVersion.get_pulp_type()
+        )
+        content = CollectionVersion.objects.filter(pk__in=filtered)
+    else:
+        content = CollectionVersion.objects.filter(pk__in=content_hrefs)
+    signing_service = SigningService.objects.get(pk=signing_service_href)
+    first_stage = CollectionSigningFirstStage(content, signing_service)
+    SigningDeclarativeVersion(first_stage, repository).create()
+
+
+class SigningDeclarativeVersion(DeclarativeVersion):
+    """Custom signature pipeline."""
+
+    def pipeline_stages(self, new_version):
+        """The stages for the signing process."""
+        pipeline = [
+            self.first_stage,  # CollectionSigningFirstStage
+            ContentSaver(),
+        ]
+        return pipeline
+
+
+class CollectionSigningFirstStage(Stage):
+    """
+    This stage signs the content and creates CollectionVersionSignatures.
+    """
+
+    def __init__(self, content, signing_service):
+        """Initialize Signing first stage."""
+        super().__init__()
+        self.content = content
+        self.signing_service = signing_service
+        # Maybe make this value a pulp_ansible constant setting?
+        self.semaphore = asyncio.Semaphore(10)
+
+    async def sign_collection_version(self, collection_version):
+        """Signs the collection version."""
+        # Limits the number of subprocesses spawned/running at one time
+        # Should everything be under the semaphore or just the signing part?
+        async with self.semaphore:
+            # We use the manifest to create the signature
+            with tempfile.TemporaryDirectory() as d:
```

Comment:
```suggestion
            with tempfile.TemporaryDirectory(dir=".") as d:
```
BTW: aiofiles.tempfile may be interesting.